pulsar-client-api/src/main/java/org/apache/pulsar/client/api/PulsarClientException.java

```diff
@@ -83,6 +86,49 @@ public PulsarClientException(String msg, Throwable t) {
         super(msg, t);
     }
 
+    /**
+     * Add a list of previous exception which occurred for the same operation
+     * and have been retried.
+     *
+     * @param previous A collection of throwables that triggered retries
+     */
+    public void setPreviousExceptions(Collection<Throwable> previous) {
+        this.previous = previous;
+    }
+
+    /**
+     * Get the collection of previous exceptions which have caused retries
+     * for this operation.
+     *
+     * @return a collection of exception, ordered as they occurred
+     */
+    public Collection<Throwable> getPreviousExceptions() {
+        return this.previous;
+    }
+
+    @Override
+    public String toString() {
```

Comment:
Why do we need to print out the previous encountered exceptions every time we log an exception?  We already log every exception in the client, can't we just search the logs for the history?

Reply:
@jerrypeng we don't print it when we log. the previous exceptions only get attached when the exception is propagated to the client. It's useful because it gives you more info to correlate on the broker side.

Reply:
@ivankelly so when will be print out the previous exceptions?

Reply:
@ivankelly aren't we printing lookup errors here:
https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConnectionHandler.java#L81

Reply:
Looking through the code, we do log exceptions returned from the broker.

Reply:
What I meant is, we don't print the exception with the previous exceptions attached every time we log. We only print that when at the point that we're about to complete the subscribeFuture or producerCreatedFuture with an exception. Which is when the exception gets passed to the client. For me, the logging of that exception is incidental. What I want is for the client code to get an exception that has context about the retries. 

Take for example the case of a customer who has a flink pipeline, and they get a TooManyRequestsException. They take a screenshot of the exception in the flink dashboard and send it to us. I want all the information to be in that screenshot, and not have to ask them to dig around in flink logs to get it.